Rebuild site with NextJS

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/doc.silverstripe.org/issues/347

Given the size of the pull-request it's probably easiest to look at the files on your local rather than in github

GuySartorelli commented on .github/workflows/keepalive.yml:
Workflows should be kept.

GuySartorelli commented on LICENSE:
..... we need to keep this lol

GuySartorelli commented on README.md:
```diff
@@ -1,135 +1,277 @@
 
-When **creating a new major branch for a pre-release major version**
+# Production build & test
+npm run build          # Build & lint (default: docs context, alias for build:docs)
```
```suggestion
npm run build          # Build, lint, and test (default: docs context, alias for build:docs)
```
If it doesn't currently run the tests, it should be updated to run the tests, consistent with every other `build` js script we run.

GuySartorelli commented on README.md:
````diff
@@ -1,135 +1,277 @@
 
-* Clone this repository to an empty directory
+```bash
+# Algolia DocSearch (optional, search disabled if not set)
````
Since this is optional, remove the "Optional Environment Variables" heading - all of the environment variables are optional it sounds like.

GuySartorelli commented on AGENTS.md:
This is for guiding coding agents for development according to that website (though I did only quickly glance at it).
Whether someone chooses to use a coding agent or not is up to the individual developer, and is the responsibility of said developer to maintain their own workflow.

Since I don't use agentic AI for my development, I don't want to review much less maintain this file.

GuySartorelli commented on README.md:
```diff
@@ -1,135 +1,180 @@
+# Testing
+npm test               # Run all tests
+npm run lint           # Run ESLint (must have 0 errors, 0 warnings)
+npm run start          # Serve built static files from out/ directory
```
What does starting the dev server (I assume that's what you mean by serving the files) have to do with testing?

GuySartorelli commented on README.md:
```diff
@@ -1,135 +1,277 @@
 
-## Building
+- **`.next/`** - Next.js build cache (dev server)
+- **`out/`** - Static HTML export (production builds via `npm run build`), served at root URL
```
It's not the most standard name for _us_ though, which I'd argue is more important.

GuySartorelli reviewed (changes requested):
I haven't looked at any of the code yet and only gave a quick click around. Here's what I've found so far:

1. The link in this block isn't obvious that it's a link:
    <img width="979" height="165" alt="image" src="https://github.com/user-attachments/assets/2defbb03-1538-4347-9287-b64215ee1c7e" />
    Same with the red variants - presumably also the same for the yellow variant but we haven't cut v7 yet so I can't tell.

1. The tab title is missing context (this PR is the top tab, existing docs are the bottom tab)
    <img width="327" height="85" alt="image" src="https://github.com/user-attachments/assets/3979bd35-e055-4215-9b21-63374e94d896" />

1. API docs aren't being rewritten correctly:
    <img width="986" height="63" alt="image" src="https://github.com/user-attachments/assets/0200cdb4-4a5e-480e-8dc9-fbfa4756d718" />

1. What's this thing and why isn't it mentioned in the readme? I assume it's only in dev mode:
    <img width="329" height="282" alt="image" src="https://github.com/user-attachments/assets/c4793cf4-1148-42db-a6e5-190de8e372b0" />

1. When in light mode the toggle looks too much like a settings cog:
    <img width="61" height="46" alt="image" src="https://github.com/user-attachments/assets/f4219fc6-cdf9-4f22-a03c-0192fb5008a9" />
    Maybe turn into a slider toggle instead? That way we see both icons at all times so it's clearer that it's light vs dark mode.

1. Mobile menu icon looks weird but that's a nitpick
    <img width="63" height="48" alt="image" src="https://github.com/user-attachments/assets/4e547f01-f729-4561-83d5-5f496afbcca3" /><img width="63" height="48" alt="image" src="https://github.com/user-attachments/assets/de58507d-f0a4-49a4-95f2-4993f9bbee56" />

1. We need a way to control the "On this page" table of contents per page or per section or something - i.e. look at the 6.0.0 changelog table of contents! It's waaaaay too long.

1. There should be no horizontal scrolling. On the 6.0.0 changelog page I have a horizontal scrollbar.

1. The "Included module versions" expandable section seems to not be rendered in changelogs.
    For that matter, are you doing something to remove table of contents from docs? This code shouldn't be choosing what to render or not render, we should remove the ToC from the docs source markdown instead.

1. TOTP auth shouldn't be directly in the "optional features" section - it should be buried under MFA.

emteknetnz commented on README.md:
```diff
@@ -1,135 +1,180 @@
-- Update `src/pages/index.tsx` to navigate to the new stable major
-- Add the new major to `redirects` in `gatsby-node.js`
+# Or combine in one command:
+npm run dev:docs      # clone + dev (happens automatically)
```
Opps, removed

GuySartorelli commented on README.md:
````diff
@@ -1,135 +1,177 @@
+### Full Command Reference
+
+```bash
+npm install            # Install dependencies (use npm ci, not npm update)
````
```suggestion
npm install            # Install dependencies (use npm install, not npm update)
```

GuySartorelli reviewed (changes requested):
Still haven't looked at the code itself, this is just from more clicking around:

1. The light/dark toggle is still a bit unclear - this time because it's not obvious which item is currently active.
    <img width="68" height="45" alt="current toggle" src="https://github.com/user-attachments/assets/67055b9a-2383-426c-8968-c1da642ade6f" />
    Something more like this is what I was thinking - with icons replacing the text. The slider makes it clear which is active currently.
    <img width="93" height="35" alt="example slider toggle" src="https://github.com/user-attachments/assets/840cf10f-da12-48a4-af20-e9e1aba91d9d" />
1. Links in "On this page" aren't visually obvious that they're links. Should just use the same styles as all other links.
    <img width="400" height="196" alt="on this page" src="https://github.com/user-attachments/assets/023c5e67-232e-40ea-834f-f45ee4e49a5c" />
1. Using `npm run build:docs && npm run start`, changelog links don't work (they work fine with `npm run dev:docs` - avoiding this sort of problem is one of the reasons I didn't want a separate "dev" mode that builds differently to "prod" mode). For example, clicking on the "5.4.0" changelog link took me to `http://localhost:3000/en/5/changelogs/5.4.0/` which shows this:
    <img width="1099" height="258" alt="weird page" src="https://github.com/user-attachments/assets/13067b91-f721-4690-99da-a13ad0f4ef54" />
1. I still have horizontal scrolling on any width between 1024px and 1250px, and anything less than 916px.


## From previous feedback

>> We need a way to control the "On this page" table of contents per page or per section or something - i.e. look at the 6.0.0 changelog table of contents! It's waaaaay too long.
>
> I don't agree, I think the large TOC is actually really good. The changelog's in particular are more like a reference document then anything else, and it seems particularly useful to have that large detailed TOC when doing an upgrade. It also helps just give an idea of what's actually in the very large changelog document.

I don't think the large TOC is good. Everything under "Full list of removed and changed API (by module, alphabetically)" is just going to be ignored, and gives both more visual clutter and more that people have to scroll past. The longer the TOC is, the more likely it is people will just ignore it, which we don't want.

> It also helps just give an idea of what's actually in the very large changelog document.

The top level of the TOC already provides most of the information. Beyond that IMO we need more control.

>> For that matter, are you doing something to remove table of contents from docs? This code shouldn't be choosing what to render or not render, we should remove the ToC from the docs source markdown instead.
>
> Yes there's some code in there to auto remove them. This seems like a a good thing? I don't know why we wouldn't want this. I'd rather not have to do PR's to remove the existing TOC's from all the changelogs

The docs should be rendered as they are, however they are. If there's a list at the top, that list might be TOC or it might not be - making an implementation that's robust enough to be 100% correct about it is hard. I'll happily make the PRs to remove the TOC from existing docs if making those PRs is the only reason you want to keep that code in there.

>> TOTP auth shouldn't be directly in the "optional features" section - it should be buried under MFA.
>
> This one's just too much work to be worth it honestly. Right now all the code expects to put an "optional module" on the "Optional modules" page.
>
> Right now TOTP is the only module that utilises a feature where you can configure where in the tree it shows (i.e. not just on the optional modules page) i.e. https://github.com/silverstripe/doc.silverstripe.org/blob/master/sources-docs.js#L159C9-L159C9
>
> However it looks goofy in CMS 6 docs because it's the only authenticator - https://docs.silverstripe.org/en/6/optional_features/mfa/authenticators/ - this would make more sense if we were planning to support webuthn long term but we're not.

I don't have a way to judge how much work it would be so I can't speak to if it'll be worth it without an estimate of the effort involved.
That said, if you insist on keeping it in the main list, it should be directly below MFA because it's completely useless without MFA - and the `/en/6/optional_features/mfa/authenticators/` page needs to be removed.


## Please also respond to

https://github.com/silverstripe/doc.silverstripe.org/pull/352#discussion_r2617679179

emteknetnz commented:
> 1. The light/dark toggle is still a bit unclear - this time because it's not obvious which item is currently active.

That slider you've provided looks kind of hard to implement, and has a tiny font which is hard to read. What I've done instead is retain the existing icons, made in the inactive icon grey and increased the icon size

> 2. Links in "On this page" aren't visually obvious that they're links. Should just use the same styles as all other links.

Updated

> 3. Using npm run build:docs && npm run start, changelog links don't work

Issue turned out to be the that `.` in the changelog urls e.g. `/en/6/changelogs/6.0.0` were interpreted as file extensions by `serve`. I've switched it out for `http-server` and things now work as expected

> I don't think the large TOC is good. Everything under "Full list of removed and changed API ...

I've updated the auto TOC to exclude an headings that are inside of an expander section. This significantly reduces the TOC size on changelogs

> The docs should be rendered as they are ... I'll happily make the PRs to remove the TOC from existing docs if making those PRs is the only reason you want to keep that code in there.

The regular changelogs TOC now render underneath the auto TOC.

> That said, if you insist on keeping it in the main list, it should be directly below MFA because it's completely useless without MFA - and the /en/6/optional_features/mfa/authenticators/ page needs to be removed.

The optional feature modules are all listed alphabetically, so there's no way to control where TOTP shows. However I don't think we need to alter the order they show in - MFA already says install TOTP - https://docs.silverstripe.org/en/6/optional_features/mfa/ - and silverstripe/mfa is a composer dep of silverstripe/totp-authenticator.

Seems like all we need to do is remove the "Authenticators" page from the mfa docs

> CI tests are also not running on this PR

Have added, it's now running. I've somewhat based it on the existing ci.yml files e.g. `permissions: {}`, don't run if pull-request is from same github account (I'm not sure why we do this anymore, though there would have been a reason originally)

While it's running in the PR, I'm going to assume that once it's merged then you can't just open a PR where anyone can just specify any code in ci.yml and it's run - presumably it'll just use the ci.yml on the master branch instead.

emteknetnz commented:
> There is a high severity vulnerability according to npm audit

Looks like that showed up 5 days ago. Have updated

GuySartorelli commented:
If I'm on a page (e.g. `/en/6/project_governance/repository_management/`) that doesn't exist in v3 docs, and then I select v3 in the version dropdown, it gives a 404. The 404 page has a black background (even though I'm on light mode) with dark grey text.

1. The 404 page should have the same background as every other page
2. I'm on the fence about whether 404 (this PR's behaviour) or kicking you to the homepage (existing behaviour) is correct.
3. The version in the dropdown always shows v6 when the 404 happens.

GuySartorelli commented:
If I click around the navigation, remembers which navigation sections were previously open and keeps them open.

This means there are more and more open sections the more I click around, so I have to go manually collapse them all again. This should probably only keep the section you're currently in open.

GuySartorelli commented on scripts/copy-images.mjs:
```diff
@@ -0,0 +1,154 @@
+
+function copyStaticAssets(destDir) {
+  const assetsDir = path.join(rootDir, 'assets');
+  const assetFiles = ['favicon.ico', 'logo.svg'];
```
Why hardcoded files? Shouldn't we just copy all files from the assets dir?

emteknetnz replied:
Have updated to just copy all files in dir

GuySartorelli commented on src/components/SearchBox.tsx:
```diff
@@ -1,51 +1,86 @@
+      indexName={indexName}
+      apiKey={apiKey}
+      disableUserPersonalization
+      searchParameters={{
```
According to my IDE `indexName` and `searchParameters` are deprecated. I won't hold off merging if these are left as they are though.

GuySartorelli commented on README.md:
Please add back the "What to update when creating a new pre-release major branch, making a stable major release, or making a major EOL" section and update it as appropriate.

emteknetnz replied:
Added back in

GuySartorelli reviewed (changes requested):
I saw a couple of `SilverStripe` which should be updated to `Silverstripe CMS`.

Other than that, most of the below is from a quick code review. I doubt I'll look at the code any closer than I did for this pass.

emteknetnz commented:
> Too big of a gap after the last item in a sublist, e.g the gap before "Dependency changes" in the below screenshot from /en/6/changelogs/6.0.0/

That's styling is on the old "manual" TOC, which you've said you were going to create some PR's to remove now that we've got an automatically generated TOC.

> There is a high severity vulnerability according to npm audit

Has been resolved

> The green inside collapsible areas looks a bit weird - maybe just use white instead?

Have changed to very light grey that's almost white.

> There's still a horizontal scroll on /en/6/changelogs/6.0.0/

Fixed

>> Seems like all we need to do is remove the "Authenticators" page from the mfa docs
> Please raise a PR for that.

Done - https://github.com/silverstripe/silverstripe-mfa/pull/626

> The 404 page should have the same background as every other page

Updated

> I'm on the fence about whether 404 (this PR's behaviour) or kicking you to the homepage (existing behaviour) is correct.

404 seems much more correct to me, that's the point of a 404 after all. This means the behaviour between changing the version selector to a missing page, and entering the url directly in the browser act the same

> The version in the dropdown always shows v6 when the 404 happens.

Fixed

> If I click around the navigation, remembers which navigation sections were previously open and keeps them open.

Have removed that behaviour

> I saw a couple of SilverStripe which should be updated to Silverstripe CMS.

Have updated all the instances of it that I found with a text search

emteknetnz commented on scripts/clone-docs.mjs:
```diff
@@ -0,0 +1,286 @@
+            // For user context at root level, only walk into userguide
+            continue;
+          }
+        }
```
Updated

GuySartorelli commented:
>> Too big of a gap after the last item in a sublist, e.g the gap before "Dependency changes" in the below screenshot from /en/6/changelogs/6.0.0/
>
> That's styling is on the old "manual" TOC, which you've said you were going to create some PR's to remove now that we've got an automatically generated TOC.

That TOC is just an unordered list. The same styling will apply to any unordered list (and presumably ordered lists as well), so it needs to be fixed.

GuySartorelli commented:
Children blocks have a couple of problems.

<details>
<summary>Click to see screenshots</summary>

Existing docs site:
<img width="1081" height="737" alt="image" src="https://github.com/user-attachments/assets/e5e4eba0-8553-4f0f-bcbc-31add57a89f4" />

With this PR: 
<img width="1137" height="791" alt="image" src="https://github.com/user-attachments/assets/74668775-b61e-4840-9ad8-4e90bac2a02b" />

</details>

Using `/en/6/optional_features/mfa/` as an example, there's two problems:
1. A link to `/en/6/optional_features/mfa/userguide/` which should be ommitted (this should be described in the sources json file)
2. The order doesn't match the existing docs site

I note the sort order in the left menu also matches the children list - i.e. the sort order there isn't correct either. I think it's treating folders as always going after files, instead of relying purely on the names of files/folders to select the sort order.

GuySartorelli commented:
Child boxes also don't expand vertically when they need to, e.g
<img width="1121" height="587" alt="image" src="https://github.com/user-attachments/assets/1130cf8c-4b23-4562-8aed-a668b7f7d89f" />
In this example from `/en/6/optional_features/` the "TinyMCE HTML editor" box should be the same height as the Linkfield box.

GuySartorelli reviewed (changes requested):
If I'm in a section that's too deeply nested to appear in the side menu, the side menu completely collapses instead of showing keeping the sections open that would lead here.

For example `/en/6/developer_guides/templates/how_tos/navigation_menu/` should keep "Developer Guides > Templates And Views" expanded.

The breadcrumb also seems to exclude the current page - the last item in the breadcrumb on this page is "How To's" which is the parent page.

GuySartorelli commented:
There's an indent before headings in a `<details>` collapsible block which isn't applied to lists. For example in the 6.0.0 changelog:
<img width="616" height="585" alt="image" src="https://github.com/user-attachments/assets/a4de393f-3246-4def-89d9-3fa3dda2f956" />

Either no content should be indented or all should.

GuySartorelli commented:
In `/en/6/contributing/documentation/#link-to-children` some of the content inside code blocks has accidentally been stripped out.

<details>
<summary>Click to see screenshots</summary>

Current docs site
<img width="1026" height="794" alt="image" src="https://github.com/user-attachments/assets/1e5cad53-a466-4322-b57f-fd612999e2e5" />

This PR
<img width="1081" height="838" alt="image" src="https://github.com/user-attachments/assets/2286cd35-58a4-416d-af95-24d962c173e9" />

</details>

emteknetnz commented:
> If I'm in a section that's too deeply nested to appear in the side menu, the side menu completely collapses instead of showing keeping the sections open that would lead here. ... The breadcrumb also seems to exclude the current page 

Both fixed

> There's an indent before headings in a `<details>` collapsible block which isn't applied to lists. For example in the 6.0.0 changelog:

Fixed

> In /en/6/contributing/documentation/#link-to-children some of the content inside code blocks has accidentally been stripped out.

Fixed

GuySartorelli reviewed (changes requested):
>> If I'm in a section that's too deeply nested to appear in the side menu, the side menu completely collapses instead of showing keeping the sections open that would lead here. ... The breadcrumb also seems to exclude the current page
> 
> Both fixed

You increased the number of items that can show in the left menu, but you didn't actually fix either of the underlying bugs. I tested by adding a very deeply nested item.

I think 4 (rather than 5 which seems to be what you've set it to now) is an appropriate number of items to allow in the left menu - and if we nest deeper than that, that should work also.

All other problems are fixed from what I can tell so just this last holdout.

Edit: Have split out: https://github.com/silverstripe/doc.silverstripe.org/issues/354

GuySartorelli reviewed (approved):
Ready to reploy I guess, lets see what happens :smiling_imp:

GuySartorelli commented:
Yay everything's green! But the preview doesn't include search still.

GuySartorelli reviewed (approved):
Looks good in preview.
